source/opt/aggressive_dead_code_elim_pass.cpp

```diff
@@ -241,15 +415,41 @@ bool AggressiveDCEPass::AggressiveDCE(ir::Function* func) {
     if (KillInstIfTargetDead(&ai))
       modified = true;
   }
-  // Kill dead instructions
-  for (auto& blk : *func) {
-    for (auto& inst : blk) {
-      if (dead_insts_.find(&inst) == dead_insts_.end())
+  // Kill dead instructions and remember dead blocks
+  std::unordered_set<ir::BasicBlock*> dead_blocks;
+  for (auto bi = structuredOrder.begin(); bi != structuredOrder.end();) {
+    uint32_t mergeBlockId = 0;
+    for (auto ii = (*bi)->begin(); ii != (*bi)->end(); ++ii) {
+      if (dead_insts_.find(&*ii) == dead_insts_.end())
         continue;
-      def_use_mgr_->KillInst(&inst);
+      // If dead instruction is selection merge, remember merge block
+      // for new branch at end of block
+      if (ii->opcode() == SpvOpSelectionMerge)
+        mergeBlockId = 
+            ii->GetSingleWordInOperand(kSelectionMergeMergeBlockIdInIdx);
+      def_use_mgr_->KillInst(&*ii);
       modified = true;
     }
+    // If a structured if was deleted, add a branch to its merge block,
+    // and traverse to the merge block killing all instructions on the way
+    // and remembering dead blocks for later deletion from function.
+    if (mergeBlockId != 0) {
+      AddBranch(mergeBlockId, *bi);
```

Comment:
How are we guaranteed that the block still exists at all?   I think this is actually ok because the block's instruction iteration skips over the OpLabel at the start. But it's worth a comment.

Is there a test case where the entire function is dead?  That would be a nice illustration that this aspect works.